### PR TITLE
Refactor code and add pagination in MCF Browser

### PR DESCRIPTION
**Pull request by cwcrystal8**
This PR has the following changes:

- updated version of react-scripts
- move graph explorer code into its own class
- add pagination to graph explorer with a new component
- adding back in a "return" value that was mistakenly deleted in an earlier PR

All the relevant changes are in package.json and the src/* files (other changes are auto-generated).

### CODE REVIEW

**Comment by cwcrystal8** [comment]
Made a slight change to where methods are stored, PR is ready for review now

**Review by pradh — COMMENTED** [review]
Not just a refactor, since it looks like this is also adding paging functionality?  That's ok, but good to call it out clearly.

Jennifer or Bo should probably have a look and approve.  LGTM.

File: mcf-browser/src/GraphExplorer.tsx [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    * IDs for nodes stored in App's state which are the subject nodes of
+    * triples from any parsed files.
+    */
+   subjNodes: string[];
+
+   /**
+    * Set id parameter in url to the given id. Used when user clicks a
+    * subject node to explore.
+    */
+   goToId: Function;
```
Comment (chejennifer) [inline mcf-browser/src/GraphExplorer.tsx:40]: nit: usually define the type for a function like `(dcid: string) => void` 
nit: maybe `onNodeClick` or something

File: mcf-browser/src/GraphExplorer.tsx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+   /** Tracks which page the user is currently viewing (0-indexed) */
+   page: number;
+ }
+
+/** Component to display the Graph explorer */
+class GraphExplorer extends Component<
+   GraphExplorerPropType,
+   GraphExplorerStateType
+ > {
+  /** Constructor for class, sets initial state
```
Comment (chejennifer) [inline mcf-browser/src/GraphExplorer.tsx:53]: nit: usually don't need function comments for constructor and render 

Here and everywhere else

File: mcf-browser/src/PageBar.tsx [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+     /** The function to go to a certain page */
+     goToPage: Function;
+ }
+
+/** Component for rendering the pagination bar */
+class PageBar extends Component<PageBarPropType> {
+  /** Constructor for the component
+        * @param {PageBarPropType} props the props passed in
+       */
+  constructor(props: PageBarPropType) {
```
Comment (chejennifer) [inline mcf-browser/src/PageBar.tsx:41]: if there is no state, the constructor actually isn't needed!

File: mcf-browser/src/GraphExplorer.tsx [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    super(props);
+    this.state = {
+      page: 0,
+    };
+  }
+
+  /** Renders the GraphExplorer component.
+    * @return {Object} the component using TSX code
+    */
+  render() {
```
Comment (chejennifer) [inline mcf-browser/src/GraphExplorer.tsx:66]: besides for the constructor, should include return type for functions (`render(): JSX.Element {...}`)

Here and everywhere else

**Review by chejennifer — COMMENTED** [review]
Thanks for refactoring!

File: mcf-browser/src/GraphExplorer.tsx (hunk 2 above)
Comment (cwcrystal8) [reply mcf-browser/src/GraphExplorer.tsx:53]: Oh the linter throws an error if I don't add a comment and the version of react I'm using won't start a server unless it passes the linter. I could disable the linter for the constructor/render functions but that seems potentially dangerous

Comment (chejennifer) [reply mcf-browser/src/GraphExplorer.tsx:53]: oh, interesting, this is fine to keep then

File: mcf-browser/src/GraphExplorer.tsx (hunk 1 above)
Comment (cwcrystal8) [reply mcf-browser/src/GraphExplorer.tsx:40]: done!

File: mcf-browser/src/PageBar.tsx [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+ interface PageBarPropType {
+     /** Tracks which page the user is currently viewing (0-indexed) */
+     page: number;
+
+     /** Stores the maximum number of pages */
+     maxPage: number;
+
+     /** The function to go to the next page */
+     goToNextPage: Function
```
[the remark this answers is not in the document]
Comment (cwcrystal8) [reply mcf-browser/src/PageBar.tsx:27]: done

File: mcf-browser/src/PageBar.tsx (hunk 3 above)
Comment (cwcrystal8) [reply mcf-browser/src/PageBar.tsx:41]: done!

File: mcf-browser/src/PageBar.tsx [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+          onClick={() => this.props.goToPage(parseFloat(pageString))}
+        >
+          {pageString}
+        </span>
+      );
+    }
+  }
+
+  /** Get page strings given the current page
+    * @param {number} page the current page that the user is on
```
[the remark this answers is not in the document]
Comment (cwcrystal8) [reply mcf-browser/src/PageBar.tsx:74]: I think JSDoc syntax has 
@param {<type>} <paramName> <description>

so in this case the "page" after "{number}" is the param name

File: mcf-browser/src/PageBar.tsx [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    * @return {string[]} the strings to render in the pagination bar
+    */
+  getPageStrings(page: number) {
+    // Get all page strings
+    const actualPage = page + 1; // Switch from 0-indexed to 1-indexed
+    const maxPage = this.props.maxPage;
+    let pagesToDisplay;
```
[the remark this answers is not in the document]
Comment (cwcrystal8) [reply mcf-browser/src/PageBar.tsx:81]: done!

File: mcf-browser/src/index.css [hunk 8]
```diff
@@ -251,3 +251,39 @@ code {
...
+  align-items: center;
+  justify-content: center;
+}
+
+.page-box {
+  border: solid 1px #ffffff;
+  border-radius: 10px;
+  height: 50px;
+  padding: 5px 20px;
+  /* width: 50px; */
```
[the remark this answers is not in the document]
Comment (cwcrystal8) [reply mcf-browser/src/index.css:277]: Oh good catch, this should be deleted.

File: mcf-browser/src/PageBar.tsx [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+    return pageStrings;
+  }
+
+    /** Returns previous page
+    * @param {number} currPage the current page the user is on
+    * @return {number} the previous page
+    */
+  static getPrevPage(currPage: number) {
+    const newPage = (currPage === 0) ? 0 : currPage - 1;
```
[the remark this answers is not in the document]
Comment (cwcrystal8) [reply mcf-browser/src/PageBar.tsx:116]: changed!

File: mcf-browser/src/PageBar.tsx [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    * @return {string[]} the strings to render in the pagination bar
+    */
+  getPageStrings(page: number) {
+    // Get all page strings
+    const actualPage = page + 1; // Switch from 0-indexed to 1-indexed
+    const maxPage = this.props.maxPage;
+    let pagesToDisplay;
+
+    // Check if page is on the boundary
+    if (actualPage < 3 || actualPage > maxPage - 2) {
```
[the remark this answers is not in the document]
Comment (cwcrystal8) [reply mcf-browser/src/PageBar.tsx:84]: Yep they're how many page strings should be shown on the boundary, changed it to a constant!

File: mcf-browser/src/GraphExplorer.tsx (hunk 4 above)
Comment (cwcrystal8) [reply mcf-browser/src/GraphExplorer.tsx:66]: Updated for all classes!

**Comment by cwcrystal8** [comment]
@chejennifer Thanks for your comments, the PR is ready for another review! Latest commits includes all of the changes from your previous comments except for the ones I haven't resolved.

**Review by chejennifer — APPROVED** [review]
Thanks for the changes!

File: mcf-browser/src/PageBar.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (chejennifer) [reply mcf-browser/src/PageBar.tsx:70]: ah ok, I missed that "page" was there, sounds good!

**Pull request merged**